WSUCptSCapstone-Fall2022Spring2023/psd-gamifiedapp
# Sprint2/saving and loading

closes #16 

Demo Video Link: https://youtu.be/bTIZDHmrC8A

Overview:

This pull request handles the addition of progress states to the data stored by level nodes. It then serializes this data at various points to autosave and load the player's progress through the game within the target device's persistent data directory.

Details:

This pull request primarily affects the following files:

LevelDescription.cs: Added a level ID to each level for identification, also added data storage for the level's progress record as well as a list of its prerequisite levels.

ProgressRecord.cs: An object stored by levels which represents whether they are complete, and which of their phases are complete. It is designed to be amenable to direct serialization into JSON.

SaveAndLoad.cs: Actually handles the saving and loading logic, initializing progress structures from phases and levels found in the scene, before loading in autosave data into those structures. Also handles saving autosaves when applicable.

EscapeListener.cs: Allows the user to press escape to go from the level player back to the level select, as well as to exit the game from level select. Will later be updated to trigger the opening of an escape menu.

CheckboxBuilder.cs: Dynamically builds the checkbox status indicators in the level select screen to indicate level progress.

Testing:

The functionality for this pull request was tested using the following manual FT procedure.

1. Delete autosave
2. Start Project
3. Confirm no existing level progress
4. Complete phase 1 of level 1
5. Exit game and restart
6. Confirm one phase progress remembered
7. Complete level one
8. Confirm Level 2 has been unlocked
9. Exit and restart
10. Confirm progress was saved, level 1 is complete, and level 2 is unlocked with no progress.

The current build passes this test as is.
